### ISSUE TITLE
Properly using callbacks and running custom validation

### ISSUE THREAD

**Issue by danielmorozoff** [ultralytics/ultralytics#20524]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi i am looking for some help on properly hooking into the training loop and running a custom validator. Everything seems to be working but the model gets put into `eval()` mode and I cant seem to put it back into `train()`

```python

    def on_train_epoch_end(trainer):
        """Called automatically at the end of each training epoch."""
        print(f"\nRunning Validation {trainer.epoch}...")

        val_res = model.val(
            data=dataset_yaml_path,
            imgsz=img_size,
            batch=batch_size,
            max_det=20000,
            workers= workers,
            device=device
        )
        print_validation(val_res)
       # HOW DO I SET THE MODEL TO TRAIN HERE?


     # Initialize and train model
    model.add_callback("on_train_epoch_end", on_train_epoch_end)
    results = model.train(
        data=dataset_yaml_path,
        epochs=epochs,
        batch=batch_size,
        imgsz=img_size,
)
```

when this runs, i get a gradient error after validation pass executes

```
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass...
```

I have tried setting the `trainer.model` object to train as well as the `trainer.model.model` object to train as well as `model.train()`

Any help would be much appreciated !

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @danielmorozoff, thank you for reaching out and for providing detailed context and code snippets! 🚀 This is an automated response to help guide your issue, and an Ultralytics engineer will also assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for additional guidance on using callbacks and custom validation. There are many helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples available.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible, as this will help us debug the issue more efficiently.

For custom training or callback usage questions, sharing complete details—including dataset samples, training logs, and steps to reproduce—helps us provide more targeted support. Please also review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for real-time discussion and support:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse](https://community.ultralytics.com/)
- [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Make sure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This ensures your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following verified, up-to-date environments (with all dependencies such as [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience while we review your question!

**Comment by RizwanMunawar** [comment]
@danielmorozoff Ok, I got your point. You want to validate the model every epoch with the callbacks? If that's the case, you can use the code mentioned below.

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

batch_size = 2
imgsz = 640
workers = 1
epochs = 3
data = "coco8.yaml"
max_det = 20000
device = "0"

def on_train_epoch_end(trainer):
    print(f"\nRunning Validation {trainer.epoch}...")

    from pathlib import Path
    val_model = YOLO(Path(trainer.save_dir) / "weights" / "best.pt")

    validation_results = val_model.val(
        data=data,
        imgsz=imgsz,
        batch=batch_size * 2,  # validation with 2x batch required
        max_det=max_det,
        workers=workers,
        device=device
    )
    # print(validation_results)

if __name__ == "__main__":
    # Add on_train_epoch_end callback.
    model.add_callback("on_model_save",
                       on_train_epoch_end)

    # Run model training on custom dataset.
    results = model.train(data=data,
                          epochs=epochs,
                          batch=batch_size,
                          imgsz=imgsz,
                          device=device,
                          workers=workers)

```

In your code, you're using the same model instance for both training and validation, which won't work as intended. Even if you manage to make it run, the results will be misleading because the validation will be performed using the pretrained YOLO model rather than the custom-trained one. What you want is to validate using the updated model checkpoints trained on your dataset.

Please let me know If you have any questions. Thanks

**Comment by danielmorozoff** [comment]
@RizwanMunawar Thank you -- will give this a go.
